executor: Support enable and disable double fetch detection

Merna177 opened the pull request:
Starting to enable/disable DF_DETECTION according to the program test

ramosian-glider commented on executor/executor_linux.h:
```diff
@@ -259,3 +262,21 @@ static void setup_machine()
+	int fd = open("/sys/kernel/debug/df_detection", O_RDWR);
+	if (fd == -1)
+		fail("open of /sys/kernel/debug/df_detection failed");
+	if (dup2(fd, dfetch->fd) < 0)
```
I'm not really sure what is going on here. In most cases dfetch->fd is initially zero, passing it by value to dup2() will probably just close fd 0 and dup the new descriptor into it. Because many programs are going to continue using fd 0, that'll be a mess.
I think once you've checked that `fd != -1` you can just assign `fd` to `dfetch->fd`.

ramosian-glider replied:
As discussed yesterday, this code actually makes sense in the case where dfetch->fd is initialized with a predefined constant, like this is done in executor.cc

On the other hand, we don't really need to keep this descriptor around. Instead, we could make `df_enable()` a function that opens the file, calls the ioctl and closes the file.

Merna177 commented:
adding df_enable in threads loop it may not be the best or the right way but it make syzkaller detect a crash (Multi-read) it doesnt mean that it solve the problem totally but i push it to know the last results i made.

melver commented on executor/executor.cc:
```diff
@@ -1543,4 +1544,4 @@ void debug_dump_data(const char* data, int length)
 	if (i % 16 != 0)
 		debug("\n");
-}
+}
```
Please undo unnecessary newline change.

ramosian-glider replied:
@Merna177 please fix (there was a newline following this brace)

melver commented on executor/executor.cc:
```diff
@@ -421,6 +421,7 @@ int main(int argc, char** argv)
 			threads[i].cov.fd = kCoverFd + i;
 			cover_open(&threads[i].cov, false);
 			cover_protect(&threads[i].cov);
+			df_enable();
```
Why is this added here? And why is it in this loop?

I think the way to check for capabilities is to first have syzkaller check if the feature exists. As an example, have a look at how KCSAN is being enabled (search for 'setup_kcsan' and look for uses of '/sys/kernel/debug/kcsan').

ramosian-glider replied:
Right, we do need to check for capabilities, this is an intermediate solution that will enable our feature unconditionally.
The intention is enable it on a per-task basis, therefore we want to call it on every thread. I don't think it works as intended right now though.

Merna177 replied:
yeah i just add this for testing to figure out what is wrong, it not the right way as i mention in the previous comment but i push it to let you see what i am trying to test

ramosian-glider commented on executor/executor_linux.h:
```diff
@@ -263,20 +263,14 @@ static void setup_machine()
-static void df_enable(df_detection* dfetch)
-{
-	if (ioctl(dfetch->fd, DF_ENABLE))
+	if (ioctl(fd, DF_ENABLE,KCOV_TRACE_PC))
```
You don't need the third argument here.

ramosian-glider commented on executor/executor.cc:
```diff
@@ -431,6 +430,7 @@ int main(int argc, char** argv)
 		}
 	}
 
+	df_enable();
```
Covering all cases is a bit more complex.
Sandboxing code below forks the process before calling loop(), so for sandboxed programs this call won't have any effect.
I suggest moving it to execute_one(), which is already executed by the proper process.

ramosian-glider reviewed (commented):
Please remember to introduce a flag for df_enable(), so that syzkaller can turn it on and off.
Once done, please squash these patches together, so that it's easier to apply them locally.

Merna177 commented:
These changes in the previous commits for testing purpose.
1) I remove double fetch flag from config flags and added it to execFlag. ( and this part not for testing as we really need it in execFlag)
2) make a check if system call is add_key just toggle the flag.(First: this part is only for testing as we still interested to detect any double fetching bug caused by add_key syscall, i do that to know if that the right place to add the filter function in fuzzer.
Second: i am still testing it.)

Merna177 commented:
Still these commits doesn't have filtering but i am pushing every step i do.
These commits only picks Lentype arguments to start filtering process

ramosian-glider commented on prog/analysis.go:
```diff
@@ -309,3 +310,91 @@ func encodeFallbackSignal(typ, id, aux int) uint32 {
+		}
+		dfDisable := filterArguments(call, pointers, p)
+		if dfDisable {
+			return dfDisable
```
I think `return true` is a bit cleaner and more future-proof (if the condition is going to change)

ramosian-glider commented on prog/analysis.go:
```diff
@@ -309,3 +310,98 @@ func encodeFallbackSignal(typ, id, aux int) uint32 {
+	return b
+}
+
+func filterArguments(call *Call, pointers map[Arg]uint64, p *Prog) bool {
```
This function needs a comment: what are we doing, why are we doing this?

ramosian-glider commented on prog/analysis.go:
```diff
@@ -309,3 +310,98 @@ func encodeFallbackSignal(typ, id, aux int) uint32 {
+			addr -= arg.Type().UnitOffset()
+			ranges = append(ranges, pair{addr, max(arg.Size(), 1)})
+		}
+
```
This newline doesn't seem necessary (unless gofmt adds it back when you delete it)

ramosian-glider commented on prog/analysis.go:
```diff
@@ -309,3 +310,98 @@ func encodeFallbackSignal(typ, id, aux int) uint32 {
+		if elem != fields[i].Name {
+			continue
+		}
+		//TODO: Check on pointer for invalid cases
```
What do we need to resolve this TODO?

Merna177 replied:
calling some functions to check on the arg, i am now trying to understand those functions to know which are useful for this case

ramosian-glider commented on prog/analysis.go:
```diff
@@ -309,3 +310,98 @@ func encodeFallbackSignal(typ, id, aux int) uint32 {
+	return pointer
+}
+
+func DFetchAnalysis(p *Prog) bool {
```
I think DFetchAnalysis isn't the best name for this function, because it doesn't reflect the function's purpose.
You do not really perform the actual double fetch analysis here (it is done by the kernel), but rather check for overlapping arguments to decide whether or not we want to enable df analysis.
I would rather call it "WantDFetchAnalysis" (and return false if we need to disable it), "WantDisableDFetchAnalysis" (and return true in the same case), "HasOverlappingArgs" etc.

ramosian-glider replied:
You'll also need to elaborate in the comment what this function is doing.

ramosian-glider commented on sys/test/test.txt:
```diff
@@ -736,6 +736,10 @@ hint_ints {
 	int2_12	int16:12
 }
 
+# Double Fetch Analysis tests
```
Please add a period at the end of this line.

ramosian-glider commented on syz-fuzzer/proc.go:
```diff
@@ -276,10 +276,18 @@ func (proc *Proc) enqueueCallTriage(p *prog.Prog, flags ProgTypes, callIndex int
 }
 
-func (proc *Proc) executeRaw(opts *ipc.ExecOpts, p *prog.Prog, stat Stat) *ipc.ProgInfo {
+func (proc *Proc) executeRaw(opts *ipc.ExecOpts, p *prog.Prog, stat Stat, enableDF Stat, disableDF Stat) *ipc.ProgInfo {
```
I think you don't need to pass `enableDF` and `disableDF` here, as they are always the same (unlike `stat`)

ramosian-glider commented on tools/syz-execprog/execprog.go:
```diff
@@ -138,6 +138,12 @@ func (ctx *Context) run(pid int) {
 		}
 		entry := ctx.entries[idx%len(ctx.entries)]
+		dfDisable := prog.DFetchAnalysis(entry.P)
+		if dfDisable {
```
Is this output used anywhere? Can we remove it?

ramosian-glider commented on prog/filter_test.go:
```diff
@@ -0,0 +1,35 @@
+		},
+		{
+			`dfetch1(&(0x7f0000000000)={0x0, {0x0}})`,
+			"true",
```
Wait, but isn't this test supposed to return false, as we discussed yesterday?

Merna177 replied:
yeah yeah it is returning false now, i am just making some modification to the code to handle this case and i will push it with the right expected return value

Merna177 commented:
all changes in the code styling mentioned the previous commit and any new lines added are been changed by gofmt

ramosian-glider commented on syz-fuzzer/proc.go:
```diff
@@ -276,17 +276,17 @@ func (proc *Proc) enqueueCallTriage(p *prog.Prog, flags ProgTypes, callIndex int
 	dfDisable := prog.DFetchAnalysis(p)
 	if dfDisable {
-		atomic.AddUint64(&proc.fuzzer.stats[disableDF], 1)
+		atomic.AddUint64(&proc.fuzzer.stats[StatEnableDF], 1)
```
Please fix: should be StatDisableDF

ramosian-glider replied:
Or maybe the opposite...
Why are we counting this program as disabled, but set the flag?

ramosian-glider replied:
Ok, I get it. Despite the flag is called 'FlagDoubleFetch', executor treats it as 'disable double fetch detection'.

I don't think this is a good idea: by default the executor must only enable the very basic functionality, everything else should be explicitly requested via flags.
So let's change this code to pass the flag when DF detection is to be enabled, and let the executor do:
```
  if (flag_df)
    df_enable();
```

Merna177 replied:
okay i will change it

ramosian-glider commented on syz-fuzzer/proc.go:
```diff
@@ -276,17 +276,17 @@ func (proc *Proc) enqueueCallTriage(p *prog.Prog, flags ProgTypes, callIndex int
 	if dfDisable {
-		atomic.AddUint64(&proc.fuzzer.stats[disableDF], 1)
+		atomic.AddUint64(&proc.fuzzer.stats[StatEnableDF], 1)
 		opts.Flags |= (1 << 6)
```
Please use ipc.FlagDoubleFetch instead of 1<<6

Merna177 replied:
isn't i need to do something like that opts.Flags &= (^ipc.FlagDoubleFetch), as if we need it by default every thing to be true, so we will need to set here 6th bit with zero to disable DoubleFetching

ramosian-glider commented on syz-manager/manager.go:
```diff
@@ -672,7 +672,10 @@ func (mgr *Manager) saveCrash(crash *Crash) bool {
 	if crash.Type == report.MultiRead {
 		mgr.mu.Lock()
-		mgr.multiReadFrames[crash.Frame] = true
+		mgr.multiReadFrames[crash.Frame] += 1
```
What's the difference between multi-reads and memory leaks or data races? Why do we need to count them?

Merna177 replied:
from my understanding the difference that memory leaks and data races are ignored from the beginning we not even need to report any single crash for them, in case of Multi-read we want to make syzkaller report a crash for the first time this crash appears and then ignore it if it appears again

Merna177 commented:
i did some changes to filtering Arguments function and add testcases, i still need to add more testcases to check everything and most of the cases, but i just make sure that it detect any overlapped arguments if it just a pointer or a pointer with associated size argument.
I will do some test cases with target test and i will push them soon, and i will also clean the code, i just addressed comments related to test function and Function naming in filtering

Merna177 commented:
are we still need dfetch enable and disable counts in stats in dashboard?

ramosian-glider commented on pkg/report/linux.go:
```diff
@@ -28,8 +29,11 @@ type linux struct {
+	multiReadIgnores      []*regexp.Regexp
 	infoMessagesWithStack [][]byte
 	eoi                   []byte
+	multiReadMutex        sync.Mutex
```
I think we'd better move this mutex closer to multiReadIgnores.
Please also remove the empty line after it.
